buildpacktoml.go

```diff
@@ -0,0 +1,294 @@
+		if err != nil {
+			return err
+		}
+	}
+	cmd.Env = append(cmd.Env, EnvBuildpackDir+"="+b.Path)
+
+	if err := cmd.Run(); err != nil {
+		return NewLifecycleError(err, ErrTypeBuildpack)
+	}
+	return nil
+}
+
+func setupEnv(env BuildEnv, layersDir string) error {
+	if err := eachDir(layersDir, func(path string) error {
+		if !isBuild(path + ".toml") {
+			return nil
+		}
+		return env.AddRootDir(path)
+	}); err != nil {
+		return err
+	}
+
+	return eachDir(layersDir, func(path string) error {
+		if !isBuild(path + ".toml") {
+			return nil
+		}
+		if err := env.AddEnvDir(filepath.Join(path, "env")); err != nil {
+			return err
+		}
+		return env.AddEnvDir(filepath.Join(path, "env.build"))
+	})
+}
+
+func eachDir(dir string, fn func(path string) error) error {
+	files, err := ioutil.ReadDir(dir)
+	if os.IsNotExist(err) {
+		return nil
+	} else if err != nil {
+		return err
+	}
+	for _, f := range files {
+		if !f.IsDir() {
+			continue
+		}
+		if err := fn(filepath.Join(dir, f.Name())); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func isBuild(path string) bool {
+	var layerTOML struct {
+		Build bool `toml:"build"`
+	}
+	_, err := toml.DecodeFile(path, &layerTOML)
+	return err == nil && layerTOML.Build
+}
+
+func (b *DefaultBuildpackTOML) readOutputFiles(bpLayersDir, bpPlanPath string, bpPlanIn BuildpackPlan) (BuildResult, error) {
+	br := BuildResult{}
+	bpFromBpInfo := Buildpack{ID: b.Buildpack.ID, Version: b.Buildpack.Version}
+
+	// read launch.toml
+	var launchTOML LaunchTOML
+	launchPath := filepath.Join(bpLayersDir, "launch.toml")
+	if _, err := toml.DecodeFile(launchPath, &launchTOML); err != nil && !os.IsNotExist(err) {
+		return BuildResult{}, err
+	}
+
+	if api.MustParse(b.API).Compare(api.MustParse("0.5")) < 0 { // buildpack API <= 0.4
+		// read buildpack plan
+		var bpPlanOut BuildpackPlan
+		if _, err := toml.DecodeFile(bpPlanPath, &bpPlanOut); err != nil {
+			return BuildResult{}, err
+		}
+		if err := validateBOM(bpPlanOut.toBOM(), b.API); err != nil {
+			return BuildResult{}, err
+		}
+		br.BOM = withBuildpack(bpFromBpInfo, bpPlanOut.toBOM())
+		br.Met = names(bpPlanOut.Entries)
+	} else {
+		// read build.toml
+		var bpBuild BuildTOML
+		buildPath := filepath.Join(bpLayersDir, "build.toml")
+		if _, err := toml.DecodeFile(buildPath, &bpBuild); err != nil && !os.IsNotExist(err) {
+			return BuildResult{}, err
+		}
+		if err := validateBOM(launchTOML.BOM, b.API); err != nil {
+			return BuildResult{}, err
+		}
+		if err := validateBOM(bpBuild.BOM, b.API); err != nil { // TODO: maybe this validation should happen in exporter
+			return BuildResult{}, err
+		}
+		if err := validateUnmet(bpBuild.Unmet, bpPlanIn); err != nil {
+			return BuildResult{}, err
+		}
+		br.BOM = withBuildpack(bpFromBpInfo, launchTOML.BOM)
+		br.Met = names(bpPlanIn.filter(bpBuild.Unmet).Entries)
+	}
+
+	br.Labels = append([]Label{}, launchTOML.Labels...)
+	for i := range launchTOML.Processes {
```

Comment:
I _think_ this behavior changed a bit. Previously, if there was no `launch.toml` it bailed out and never set processes, slices, or labels. This iteration feels better than what was happening before, but this loop of `launchTOML .Processes` for a file that does not exist file caught my attention.

Reply:
Ah, good call... looking for a way to improve this flow.

Reply:
Latest changes have a bail out if launch.toml does not exist. It makes the function even more verbose but could probably be made more readable with some helper functions.